base/clistruct/clistruct.go

```diff
@@ -0,0 +1,152 @@
+		panic(err.Error())
+	}
+}
+
+// registerFlags is the non-panicking version of MustRegisterFlags, for testing.
+func registerFlags(fs *flag.FlagSet, cliStruct interface{}) error {
+	if fs == nil {
+		return errors.New("expected flag.FlagSet in RegisterFlags but got nil")
+	}
+
+	refVal, err := reflectStructValue(cliStruct)
+	if err != nil {
+		return err
+	}
+
+	if err := register(fs, refVal, ""); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// register recursively assigns flags to any fields that have a cli struct tag, are addressable, and have a supported type.
+func register(fs *flag.FlagSet, val reflect.Value, namePrefix string) error {
+	for i := 0; i < val.Type().NumField(); i++ {
+		field := val.Type().Field(i)
+
+		flagName, ok := field.Tag.Lookup(cliStructNameTag)
+		if !ok {
+			// skip untagged fields
+			continue
+		}
+
+		// Embed struct types without nesting when `cli:""`
+		if flagName == "" {
+			if field.Type.Kind() != reflect.Struct {
+				return fmt.Errorf("field %q can't embed for a non-struct type: %s", field.Name, field.Type.Kind())
+			}
+
+			// recurse for embedded struct
+			if err := register(fs, val.Field(i), namePrefix); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// nest the given flagName under the parent tags
+		if namePrefix != "" {
+			flagName = namePrefix + "." + flagName
+		}
+
+		fieldType := field.Type
+		fieldValue := val.Field(i)
+
+		switch fieldType.Kind() {
+		case reflect.Struct:
+			// recurse for nested structs
+			if err := register(fs, fieldValue, flagName); err != nil {
+				return err
+			}
+			continue
+		case reflect.Ptr, reflect.Interface:
+			// dereference and see if we can set it below, but it's unlikely if the given struct has not had values assigned to it.
+			fieldValue = fieldValue.Elem()
+		}
+
+		if !fieldValue.CanSet() {
+			return fmt.Errorf("field %q can't be set - not addressable or unexported", field.Name)
+		}
+
+		flagHelp, _ := field.Tag.Lookup(cliStructHelpTag)
+		if err := assignFlagToStructFieldVar(fs, fieldValue, flagName, flagHelp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// assignFlagToStructFieldVar creates a flag associated with the given struct field, if the type can be set from a flag.
+func assignFlagToStructFieldVar(fs *flag.FlagSet, fieldValue reflect.Value, flagName, flagHelp string) error {
+
+	switch v := reflectInterfaceValue(fieldValue).(type) {
+	case flag.Value:
+		fs.Var(v, flagName, flagHelp)
+	case *time.Duration:
+		fs.DurationVar(v, flagName, 0, flagHelp)
+	case *json.Number:
+		fs.Var((*jsonNumberFlagValue)(v), flagName, flagHelp)
+	default:
+		// See docs: unsafe.Pointer (5) Conversion of the result of ... reflect.Value.UnsafeAddr
+		// uintptr -> unsafe.Pointer conversion MUST be done without an intermediate variable
+		p := unsafe.Pointer(fieldValue.UnsafeAddr())
+
+		// we can only map struct field types which have a matching flag var type
+		switch fieldValue.Type().Kind() {
+		case reflect.Bool:
```

Comment:
What's the difference between what we're doing here, and the handling in https://github.com/golang/tools/blob/v0.1.0/internal/tool/tool.go#L187 to generate flags for a reflect.Value?  The latter looks like it's handling these types directly in the type switch above.

Reply:
Huh, that looks way better, seems easy to get rid of this reflect Kind switch (and unsafe in the process) - which makes me more comfortable putting this directly in `base`.